File: src/advancedSort/mergeSort.js
function merge(array){
    if (array.length <= 1) {
        return array;
    }
    const middle = Math.floor(array.length / 2);
    let left = array.slice(0, middle);
    let right = array.slice(middle, array.length);

    left = merge(left);
    right = merge(right);

    return mergeAll(left, right);
}

function mergeAll(left,right) {
    let sorted = [];
    let i = 0 , j = 0;
    
    while (i < left.length && j < right.length) {
        if (left[i] < right[j]) {
            sorted.push(left[i]);
            i++;
        } else {
            sorted.push(right[j]);
            j++;
        }
    }
    return sorted.concat(left.slice(i) , right.slice(j));
}


arr = [1543, 2341, 7890, 5678, 4321, 9087, 3210, 8765, 4567, 1098, 5432, 9876, 8765, 2310, 5678, 9087, 8765, 1098, 6543, 3210]
sortedArray = merge(arr);

console.log("\nTampilan Awal Array: ");
console.log(arr)

console.log("");

console.log("Tampilan Akhir Setelah Dilakukan Merge Sorting : ");
console.log(sortedArray)

File: src/advancedSort/mergeSort.java
package advancedSort;

public class mergeSort {
    public static void main(String[] args) {
        int[] Arrs = {1,3,4,5,2,4,8,9,10};

        for (int i = 0; i < Arrs.length; i++) {
            System.out.print(Arrs[i]+ " ");
        }
        merge(Arrs);
        System.out.println(" ");
        for (int i = 0; i < Arrs.length; i++) {
            System.out.print(Arrs[i]+" ");
        }
    }
    private static void merge(int[] arr){
        
        if (arr.length <= 1) {
            return;
        }
        int middle = arr.length/2;
        int[] left = new int[middle];
        int[] right = new int[arr.length - middle];

        int i = 0;
        int j = 0;
        for (;i < arr.length; i++) {
            if (i< middle) {
                left[i] = arr[i];
            }else{
                right[j] = arr[i];
                j++;
            }
        }
        merge(left);
        merge(right);
        allMerge(left, right, arr);
    }
    private static void allMerge(int[] left,int[] right,int[] arr){
        int leftSize = arr.length/2;
        int rightSize = arr.length - leftSize;
        int i= 0, l = 0 ,r =0;

        while (l < leftSize && r < rightSize) {
            if (left[l] < right[r]) {
                arr[i] = left[l];
                i++;
                l++;
            }else{
                arr[i] = right[r];
                i++;
                r++;
            }
        }
        while (l < leftSize) {
            arr[i] = left[l];
            l++;
            i++;
        }
        while (r < rightSize) {
            arr[i] = right[r];
            r++;
            i++;
        }
    }
}
